### ISSUE TITLE
Macro __cplusplus seems not working by default (MSVC)

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#314]
On behalf of @sgf:
about __cplusplus seems do not work default on MSVC (Need Add /Zc:__cplusplus or /Zc:__cplusplus- to the Additional options: pane. )
or
#if ((defined(_MSVC_LANG) && _MSVC_LANG >= 201703L) || __cplusplus >= 201703L)
//C++17 specific stuff here
#endif

**Comment by sgf** [comment]
Thanks you feedback for my report.


[https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019)

_MSC_VER from (VC6)
_MSVC_LANG from VS2015Update3

about __cplusplus
[https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=vs-2019](https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=vs-2019)

**Comment by mojca** [comment]
Wouldn't it make sense to add `/Zc:__cplusplus` to CMakeLists?

**Comment by ivmai** [comment]
> Wouldn't it make sense to add `/Zc:__cplusplus` to `CMakeLists.txt`?

It is possible, but in this case we should add it to NT_MAKEFILE.

I prefer to fix issues in the source code when possible.

@sgf, could you please make a draft patch.

**Comment by sgf** [comment]
> @sgf, could you please make a draft patch.

I am a newbie in C ++ and I have not even successfully compiled bdwgc on windows, but I will try it.

draft :

```C++
#ifndef GC_CPP_VER
#define _TMP_GC_CPP_VER  199711L
	#if(defined(_MSC_VER) && _MSC_VER>=1200) //msvc 6.0+
		#ifdef __cplusplus
			#if (_MSVC_LANG==__cplusplus) // /Zc:__cplusplus opened
				#define _TMP_GC_CPP_VER __cplusplus
			#else // /Zc:__cplusplus closed(default is closed)
				#define _TMP_GC_CPP_VER _MSVC_LANG
			#endif
		#elif defined(_MSVC_LANG)//NO __cplusplus BUT _MSVC_LANG
			#define _TMP_GC_CPP_VER _MSVC_LANG
		#else NO __cplusplus NO _MSVC_LANG
			#define _TMP_GC_CPP_VER 199711L
		#endif
		#if (_TMP_GC_CPP_VER>=201704L)// /std:c++latest //C++ latest drafts,new C++ 20 maybe 202002L or others
			#define GC_CPP_VER 20 //
		#elif (_TMP_GC_CPP_VER>=201703L)
			#define GC_CPP_VER 17
		#elif (_TMP_GC_CPP_VER>=201402L)
			#define GC_CPP_VER 14
		#else //the MSVC compiler does not, and never will, support a C++11, C++03, or C++98 standards
			#define GC_CPP_VER 98
		#endif //  _MSC_VER
	#endif
	//for other C++ compiler
#endif
```

i have no idea for other  C++ compiler

**Comment by ivmai** [comment]
The /Zc:__cplusplus option is available starting in Visual Studio 2017 version 15.7.

So, to use this option, we need a VS version detection to be added to NT_MAKEFILE and CMakefile first.
